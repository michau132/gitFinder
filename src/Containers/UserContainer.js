import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import takeUserNameAndFetchData from '../actions/fetchData';
import LoaderHOC from '../hoc/LoaderHOC';
import ErrorHOC from '../hoc/ErrorHOC';

class UserContainer extends Component {
  static propTypes = {
    fetchData: PropTypes.func.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        user: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
    user: PropTypes.shape({
      userInfo: PropTypes.shape({
        login: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  }

  constructor(props, context) {
    super(props, context);
    this.state = {};
  }

  componentDidMount() {
    const { fetchData, match, user } = this.props;
    if (user.userInfo.login.length === 0) {
      fetchData(match.params.user);
    }
  }

  componentDidUpdate(prevProps) {
    const { fetchData, match } = this.props;
    const previousUser = prevProps.match.params.user;
    const nextUser = match.params.user;
    if (previousUser !== nextUser) {
      fetchData(match.params.user);
    }
  }

  render() {
    const { render } = this.props;
    return (
      <Fragment>
        {
          render({ ...this.props })
        }
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  user: state,
});

const mapDispatchToProps = {
  fetchData: takeUserNameAndFetchData,
};

const UserContainerWithLoading = ErrorHOC(LoaderHOC(UserContainer));

export { UserContainer };

export default
withRouter(
  connect(mapStateToProps, mapDispatchToProps)(UserContainerWithLoading),
);
