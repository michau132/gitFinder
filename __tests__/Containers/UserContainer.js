import React from 'react';
import { shallow } from 'enzyme';
import { observable } from 'mobx';
import UserContainer from '../../src/Containers/UserContainer';

const store = observable({
  error: false,
  isLoading: false,
  getUserInfoAndRepos: jest.fn(),
});

describe('testing UserContainer', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = {
      match: {
        params: {
          user: 'adam',
        },
      },
      render: jest.fn(),
      store,
    };
    wrapper = shallow(<UserContainer.wrappedComponent {...props} />);
  });


  test('renders without crashing', () => {
    expect(wrapper).toBeDefined();
  });

  test('getUserInfoAndRepos is called in ComponentDidMount', () => {
    wrapper.instance().componentDidMount();
    expect(props.store.getUserInfoAndRepos).toHaveBeenCalledWith('adam');
  });

  test('getUserInfoAndRepos is called when props changed', () => {
    wrapper.setProps({
      match: {
        params: {
          user: 'otherOne',
        },
      },
    });
    expect(props.store.getUserInfoAndRepos).toHaveBeenCalledWith('otherOne');
  });
});
