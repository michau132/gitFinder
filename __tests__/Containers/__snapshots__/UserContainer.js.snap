// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing UserContainer should match snapshot 1`] = `
<UserContainer
  match={
    Object {
      "params": Object {
        "user": "adam",
      },
    }
  }
  render={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "handleFilterRepos": [Function],
            "handleHideSelectedRepos": [Function],
            "handleHideSingleRepo": [Function],
            "handleSelectAllRepos": [Function],
            "handleSelectUserRepo": [Function],
            "handleShowAllRepos": [Function],
            "informations": Object {
              "avatar_url": "https://via.placeholder.com/150",
              "email": "aoko123@gmail.com",
              "loctation": "Raszyn",
              "login": "aOko123",
              "name": "Adam Oko",
            },
            "openSelectedRepos": [Function],
            "repos": Array [
              Object {
                "description": "First repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/firstRepo",
                "id": 1,
                "name": "firstRepo",
                "stargazers_count": 0,
              },
              Object {
                "description": "Second repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/secondRepo",
                "id": 2,
                "name": "secondRepo",
                "stargazers_count": 0,
              },
            ],
            "restStore": Object {
              "allReposAreSelected": false,
              "checkIsShowAllIsDisabled": [Function],
              "countIsFounded": [Function],
              "error": null,
              "fetchData": [Function],
              "filterProjectsInput": "",
              "isEveryRepoIsSelected": [Function],
              "isFoundedCount": 0,
              "isLoading": false,
              "isOneRepoSelected": [Function],
              "isShowAllBtnDisabled": true,
              "selectedReposAreEmpty": true,
            },
          },
        ],
        Array [
          Object {
            "handleFilterRepos": [Function],
            "handleHideSelectedRepos": [Function],
            "handleHideSingleRepo": [Function],
            "handleSelectAllRepos": [Function],
            "handleSelectUserRepo": [Function],
            "handleShowAllRepos": [Function],
            "informations": Object {
              "avatar_url": "https://via.placeholder.com/150",
              "email": "aoko123@gmail.com",
              "loctation": "Raszyn",
              "login": "aOko123",
              "name": "Adam Oko",
            },
            "openSelectedRepos": [Function],
            "repos": Array [
              Object {
                "description": "First repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/firstRepo",
                "id": 1,
                "name": "firstRepo",
                "stargazers_count": 0,
              },
              Object {
                "description": "Second repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/secondRepo",
                "id": 2,
                "name": "secondRepo",
                "stargazers_count": 0,
              },
            ],
            "restStore": Object {
              "allReposAreSelected": false,
              "checkIsShowAllIsDisabled": [Function],
              "countIsFounded": [Function],
              "error": null,
              "fetchData": [Function],
              "filterProjectsInput": "",
              "isEveryRepoIsSelected": [Function],
              "isFoundedCount": 0,
              "isLoading": true,
              "isOneRepoSelected": [Function],
              "isShowAllBtnDisabled": true,
              "selectedReposAreEmpty": true,
            },
          },
        ],
        Array [
          Object {
            "handleFilterRepos": [Function],
            "handleHideSelectedRepos": [Function],
            "handleHideSingleRepo": [Function],
            "handleSelectAllRepos": [Function],
            "handleSelectUserRepo": [Function],
            "handleShowAllRepos": [Function],
            "informations": Object {
              "avatar_url": "https://via.placeholder.com/150",
              "email": "aoko123@gmail.com",
              "loctation": "Raszyn",
              "login": "aOko123",
              "name": "Adam Oko",
            },
            "openSelectedRepos": [Function],
            "repos": Array [
              Object {
                "description": "First repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/firstRepo",
                "id": 1,
                "name": "firstRepo",
                "stargazers_count": 0,
              },
              Object {
                "description": "Second repo description",
                "forks_count": 0,
                "html_url": "https://github.com/aOko123/secondRepo",
                "id": 2,
                "name": "secondRepo",
                "stargazers_count": 0,
              },
            ],
            "restStore": Object {
              "allReposAreSelected": false,
              "checkIsShowAllIsDisabled": [Function],
              "countIsFounded": [Function],
              "error": null,
              "fetchData": [Function],
              "filterProjectsInput": "",
              "isEveryRepoIsSelected": [Function],
              "isFoundedCount": 0,
              "isLoading": false,
              "isOneRepoSelected": [Function],
              "isShowAllBtnDisabled": true,
              "selectedReposAreEmpty": true,
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
/>
`;
